### PR TITLE
Bring YamlTree structures down to the Rule constructor

### PR DESCRIPTION
**Pull request by rcoh**

### CODE REVIEW

File: semgrep/semgrep/semgrep_main.py [hunk 1]
```diff
@@ -168,6 +167,7 @@ def get_config(
     # first check if user asked to generate a config
     if generate_config:
         semgrep.config_resolver.generate_config()
+        sys.exit(0)
```
Comment (brendongo) [inline semgrep/semgrep/semgrep_main.py:171]: Let's continue the trend of https://github.com/returntocorp/semgrep/pull/916 and try to limit number of places we call exit. Can probably move this up the call chain

File: semgrep/semgrep/config_resolver.py [hunk 2]
```diff
@@ -40,20 +39,27 @@
 DEFAULT_REGISTRY_KEY = "r2c"
 
 
-def manual_config(pattern: str, lang: str) -> Dict[str, Any]:
+def manual_config(pattern: str, lang: str) -> Dict[str, Optional[YamlTree]]:
     # TODO remove when using sgrep -e ... -l ... instead of this hacked config
+    pattern_tree = parse_yaml_preserve_spans(pattern, filename=None)
+    we_messed_up_span = parse_yaml_preserve_spans(
```
Comment (brendongo) [inline semgrep/semgrep/config_resolver.py]: error_span? idk about `we_messed_up_span`

File: semgrep/semgrep/semgrep_main.py (hunk 1 above)
Comment (rcoh) [reply semgrep/semgrep/semgrep_main.py:171]: This is preserving existing behavior/making mypy happy

Comment (mschwager) [reply semgrep/semgrep/semgrep_main.py:171]: Yeah, this was one of the `sys.exit` calls that was more difficult to pull out. Shouldn't be too much work - I would pull the `if generate_config` check into the `cli.cli()` function, call `config_resolver.generate_config()` from there, and do an early `return`. But no worries if that seems like too much here.

Comment (brendongo) [reply semgrep/semgrep/semgrep_main.py:171]: Oh i see. sg.

**Review by brendongo — APPROVED** [review]

**Pull request merged**